### DISCUSSION TITLE
Poll: Default config file location

### DISCUSSION THREAD

**Discussion by rgwood** [nushell/nushell#9019] [Polls]
Hi, this is a follow-up from https://github.com/nushell/nushell/discussions/8263!

We’re still trying to settle the question of where to store Nushell’s configuration files (`config.nu`, `env.nu`, etc.). We plan to respect the `XDG_CONFIG_HOME` environment variable if set, but the _default_ location for configuration files is still uncertain.

Currently, Nushell uses platform-specific folders:
- `~/.config/` on Linux
- `~/Library/Application Support/` on macOS
- `~\AppData\Roaming\` on Windows

However, we've had some requests to standardize on `~/.config/` [across all OSs](https://github.com/nushell/nushell/issues/8256).

Neither approach is obviously correct; there's a case to be made for respecting OS traditions, and a case to be made for standardizing our approach across OSs. So we're asking you, the Nushell user:

**Comment by RickCogley** [comment]
The space in the default macOS config dir causes agita.

**Comment by jmoore34** [comment]
I feel like having a consistent folder aligns with nushell's philosophy of having a consistent experience across platforms. 

Seconded on the macos location agita; I use Nushell on my work MacBook and it's a hassle to navigate to

**Comment by prmadev** [comment]
I'm a Linux user so for me .config all the way. But I can imagine that if I was on other platforms, I would probably want a consistent experience across my whole computer. 
What would ~/.config look like in windows, by the way?

**Reply by rgwood** [reply]
> What would ~/.config look like in windows, by the way?

Could you clarify what you mean by this?

**Reply by prmadev** [reply]
> > What would ~/.config look like in windows, by the way?
> 
> Could you clarify what you mean by this?

I have not used windows in years. But as I remember there was no concept of "home" in windows. Also adding '.' in the beginning of the directory/file  does not have the same effect. Where does the .config folder reside in windows?

**Reply by rgwood** [reply]
Windows users have a home directory at `C:\Users\username`. Same idea as `/home/username` on Linux or `/Users/username` on macOS.

**Comment by 0xJohnnyboy** [comment]
I like how [Wezterm](https://wezfurlong.org/wezterm/config/files.html#configuration-files) handles config files, also the doc is really clear. In the end, even if I voted the `~/.config/` option, it doesn't really matter where it goes as long as it's consistent and well documented.

**Reply by fdncred** [reply]
Yup, we like the Wezterm diagram too. We'd love to have a diagram like this on our nushell.sh website somewhere. If we could just talk someone into creating it, hint hint. 🤣

**Reply by 0xJohnnyboy** [reply]
This is relatively straightforward, here is the MermaidJS code used in Wezterm doc. Once the decision has been made for Nushell, you can easily edit the blocks :)

```js
graph TD
  X[Locate Configuration file] --> A{{--config-file CLI argument specified?}}
  A -->|Yes| B{{Can that file be loaded?}}
  B -->|Yes| C[Use it]
  B -->|No| D[Use built-in default configuration]
  A -->|No| E{{$WEZTERM_CONFIG_FILE<br/>environment set?}}
  E -->|Yes| B
  E -->|No| F{{"Running on Windows and<br/>wezterm.lua exists in same<br/>dir as wezterm.exe?<br/>(Thumb drive mode)"}}
  F -->|Yes| B
  F -->|No| H{{Is $XDG_CONFIG_HOME<br/>environment set and<br/>wezterm/wezterm.lua<br/>exists inside it?}}
  H -->|Yes| B
  J --> B
  H -->|No| K{{Does $HOME/.config/wezterm/wezterm.lua exist?}}
  K -->|Yes| B
  K -->|No| J[Use $HOME/.wezterm.lua]
```

**Comment by willemv** [comment]
For the Windows case, putting it in `%APPDATA%\Roaming` has the benefit of integrating with [Roaming User Profiles](https://learn.microsoft.com/en-us/windows/win32/shell/roaming-user-profiles)

**Reply by fdncred** [reply]
Thanks for pointing this out.

**Comment by lizclipse** [comment]
I've recently moved to using [chezmoi](https://github.com/twpayne/chezmoi) to manage my configs, and this issue has cropped up again for me, so I just wanted to throw my 2¢ into the ring and skedaddle.

I'm running on macOS, and I've found the default config dir more tedious than anything. For a while now I've had a symlink into `~/.nushell -> ~/Library/Application Support/nushell` because I have more than the standard config in there and I modify it regularly, but I've moved to having a symlink `~/Library/Application Support/nushell -> ~/.config/nushell` so that it can work with chezmoi nicely.

From my perspective, having the config in the same place across platforms makes using this sort of tooling easier, but also nu is the only tool (I think) I have that doesn't just treat macOS as another unix-like system, which is why the current location is so unexpected. Changing the default or being able to specify it (and env var might work for that, but I currently using the nu config for specifying my vars, as it's my login shell, so at least some docs on how to manage that would be nice) seems to be a good path forward, however, as long as nu keeps supporting using a symlink for the config dir, I think the issue is workable for now.

**Comment by incansvl** [comment]
For the Linux and Windows platforms, the current arrangement (~/.config, %APPDATA%\Roaming respectively) have the advantage of being consistent with platform norms, and other apps on the same platform.

I also agree with the point made by @willemv. Windows tends to expect/assume apps will follow the platform norms (even if these change over time) and may well break if you do something different. A `C:\users\username\.config` directory would need to be marked `hidden` to behave like Linux, but even so this feels like a "foreign" solution on that platform.

I can't comment on MacOS as I have never used that platform.

**Reply by cstrachan88** [reply]
Windows can assume what it wants, but I'm not sure why the idea of setting `NU_CONFIG_PATH` was rejected. As others have mentioned, Wezterm handles this the best. Either way, it seems like the overwhelming majority of users prefer the `.config/nushell` path.

I prefer `.config/nushell/config.nu` on windows, but am unable to use this location without affecting other applications through the use of `XDG_CONFIG_HOME`

**Comment by jmoore34** [comment]
A little workaround I found for reducing pain on MacOS

```
ln -s ~"/Library/Application Support/" ~/config
```
